# Imports

In [1]:
#Importing the libraries

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
#Reading the datasets
leads = pd.read_csv('Leads.csv')
leads.head()

Prospect ID  Lead Number              Lead Origin  \
0  7927b2df-8bba-4d29-b9a2-b6e0beafe620       660737                      API   
1  2a272436-5132-4136-86fa-dcc88c88f482       660728                      API   
2  8cc8c611-a219-4f35-ad23-fdfd2656bd8a       660727  Landing Page Submission   
3  0cc2df48-7cf4-4e39-9de9-19797f9b38cc       660719  Landing Page Submission   
4  3256f628-e534-4826-9d63-4a8b88782852       660681  Landing Page Submission   

      Lead Source Do Not Email Do Not Call  Converted  TotalVisits  \
0      Olark Chat           No          No          0          0.0   
1  Organic Search           No          No          0          5.0   
2  Direct Traffic           No          No          1          2.0   
3  Direct Traffic           No          No          0          1.0   
4          Google           No          No          1          2.0   

   Total Time Spent on Website  Page Views Per Visit  ...  \
0                            0                   0.0  ...   
1                          674                   2.5  ...   
2                         1532                   2.0  ...   
3                          305                   1.0  ...   
4                         1428                   1.0  ...   

  Get updates on DM Content    Lead Profile    City  \
0                        No          Select  Select   
1                        No          Select  Select   
2                        No  Potential Lead  Mumbai   
3                        No          Select  Mumbai   
4                        No          Select  Mumbai   

  Asymmetrique Activity Index Asymmetrique Profile Index  \
0                   02.Medium                  02.Medium   
1                   02.Medium                  02.Medium   
2                   02.Medium                    01.High   
3                   02.Medium                    01.High   
4                   02.Medium                    01.High   

  Asymmetrique Activity Score Asymmetrique Profile Score  \
0                        15.0                       15.0   
1                        15.0                       15.0   
2                        14.0                       20.0   
3                        13.0                       17.0   
4                        15.0                       18.0   

  I agree to pay the amount through cheque  \
0                                       No   
1                                       No   
2                                       No   
3                                       No   
4                                       No   

  A free copy of Mastering The Interview Last Notable Activity  
0                                     No              Modified  
1                                     No          Email Opened  
2                                    Yes          Email Opened  
3                                     No              Modified  
4                                     No              Modified  

[5 rows x 37 columns]

## 1. Data Understanding

In [3]:
#Reading the basic stats of the dataset without any processing
leads.describe()

Lead Number    Converted  TotalVisits  Total Time Spent on Website  \
count    9240.000000  9240.000000  9103.000000                  9240.000000   
mean   617188.435606     0.385390     3.445238                   487.698268   
std     23405.995698     0.486714     4.854853                   548.021466   
min    579533.000000     0.000000     0.000000                     0.000000   
25%    596484.500000     0.000000     1.000000                    12.000000   
50%    615479.000000     0.000000     3.000000                   248.000000   
75%    637387.250000     1.000000     5.000000                   936.000000   
max    660737.000000     1.000000   251.000000                  2272.000000   

       Page Views Per Visit  Asymmetrique Activity Score  \
count           9103.000000                  5022.000000   
mean               2.362820                    14.306252   
std                2.161418                     1.386694   
min                0.000000                     7.000000   
25%                1.000000                    14.000000   
50%                2.000000                    14.000000   
75%                3.000000                    15.000000   
max               55.000000                    18.000000   

       Asymmetrique Profile Score  
count                 5022.000000  
mean                    16.344883  
std                      1.811395  
min                     11.000000  
25%                     15.000000  
50%                     16.000000  
75%                     18.000000  
max                     20.000000

In [4]:
#shape of our data
leads.shape

(9240, 37)

In [5]:
#Datatype of each variables
leads.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9240 entries, 0 to 9239
Data columns (total 37 columns):
 #   Column                                         Non-Null Count  Dtype  
---  ------                                         --------------  -----  
 0   Prospect ID                                    9240 non-null   object 
 1   Lead Number                                    9240 non-null   int64  
 2   Lead Origin                                    9240 non-null   object 
 3   Lead Source                                    9204 non-null   object 
 4   Do Not Email                                   9240 non-null   object 
 5   Do Not Call                                    9240 non-null   object 
 6   Converted                                      9240 non-null   int64  
 7   TotalVisits                                    9103 non-null   float64
 8   Total Time Spent on Website                    9240 non-null   int64  
 9   Page Views Per Visit                           9103 

In [6]:
#TotalVisits should be converted to int
#leads['Page Views Per Visit'].value_counts() - looks weird - how can views be a decimal value

for i in leads.columns:
    print(i)
    print(leads[i].value_counts())
    print('_____________________________________')

Prospect ID
b1f85fbc-0244-4a3e-b7e3-3047478348a5    1
5a3aa671-7ee4-4ac2-824e-357471cc9d6f    1
7df9db7e-7ca4-4424-a7ea-a101f24ad5d8    1
a66b2a39-2f12-4eb4-b6e3-065d88ccf980    1
6832a282-8f10-48f8-9240-6f6a5d8a7fbc    1
                                       ..
6cc9046a-5b94-48b7-8a2e-bc3737a5d5d6    1
0c9a0616-ec5f-488a-8517-bd274216f5e3    1
1ddc803a-0ea0-415d-ad99-637bdf6e5e24    1
fc37bbbb-e1e4-4ed7-8867-496afa31fb85    1
66a14bfa-904a-4c3f-b538-83ea1b82b363    1
Name: Prospect ID, Length: 9240, dtype: int64
_____________________________________
Lead Number
608255    1
596233    1
656682    1
592824    1
652580    1
         ..
631489    1
600768    1
598556    1
616923    1
655360    1
Name: Lead Number, Length: 9240, dtype: int64
_____________________________________
Lead Origin
Landing Page Submission    4886
API                        3580
Lead Add Form               718
Lead Import                  55
Quick Add Form                1
Name: Lead Origin, dtype: int64
__________

Name: City, dtype: int64
_____________________________________
Asymmetrique Activity Index
02.Medium    3839
01.High       821
03.Low        362
Name: Asymmetrique Activity Index, dtype: int64
_____________________________________
Asymmetrique Profile Index
02.Medium    2788
01.High      2203
03.Low         31
Name: Asymmetrique Profile Index, dtype: int64
_____________________________________
Asymmetrique Activity Score
14.0    1771
15.0    1293
13.0     775
16.0     467
17.0     349
12.0     196
11.0      95
10.0      57
9.0        9
18.0       5
8.0        4
7.0        1
Name: Asymmetrique Activity Score, dtype: int64
_____________________________________
Asymmetrique Profile Score
15.0    1759
18.0    1071
16.0     599
17.0     579
20.0     308
19.0     245
14.0     226
13.0     204
12.0      22
11.0       9
Name: Asymmetrique Profile Score, dtype: int64
_____________________________________
I agree to pay the amount through cheque
No    9240
Name: I agree to pay the amount through

#### From value_counts, it can 
There are no duplicates in Prospect ID
There are no duplicates in Lead Origin

Insights:
- Most leads are sourced through Google...bla bla bla
- Most people donot want to be called or emailed for the course
- Conversion rate is { %} than not


In [ ]:
leads[leads['What matters most to you in choosing a course']=='Other'].Converted

### This is for analysis/removal of columns - single value or high concentration on one value or redundant/irrelevant columns

Update me on Supply Chain Content
No    9240
Name: Update me on Supply Chain Content, dtype: int64
_____________________________________
Get updates on DM Content
No    9240
Name: Get updates on DM Content, dtype: int64
_____________________________________
I agree to pay the amount through cheque
No    9240
Name: I agree to pay the amount through cheque, dtype: int64
_____________________________________
Search
No     9226
Yes      14
Name: Search, dtype: int64
_____________________________________
Magazine
No    9240
Name: Magazine, dtype: int64
_____________________________________
Newspaper Article
No     9238
Yes       2
Name: Newspaper Article, dtype: int64
_____________________________________
X Education Forums
No     9239
Yes       1
Name: X Education Forums, dtype: int64
_____________________________________
Newspaper
No     9239
Yes       1
Name: Newspaper, dtype: int64
_____________________________________
Digital Advertisement
No     9236
Yes       4
Name: Digital Advertisement, dtype: int64
_____________________________________
Through Recommendations
No     9233
Yes       7
Name: Through Recommendations, dtype: int64
_____________________________________
Receive More Updates About Our Courses
No    9240
Name: Receive More Updates About Our Courses, dtype: int64
_____________________________________
What matters most to you in choosing a course
Better Career Prospects      6528
Flexibility & Convenience       2
Other                           1
Name: What matters most to you in choosing a course, dtype: int64
_____________________________________
Do Not Call
No     9238
Yes       2
Name: Do Not Call, dtype: int64



### Null Value Treatment

In [ ]:
leads.columns

In [ ]:
#Checking for null values
(leads.isnull().sum() * 100)/len(leads)

In [ ]:
#We can observe null values on many columns. 
#We can drop the unwanted columns then proceed to treat the null values.

In [ ]:
# del_col = (['Newspaper','Receive More Updates About Our Courses',
#             'Update me on Supply Chain Content','Newspaper Article',
#             'I agree to pay the amount through cheque','Newspaper Article',
#            'Magazine','Get updates on DM Content','X Education Forums','Do Not Call'])

del_col = (['Newspaper','Receive More Updates About Our Courses',
            'Update me on Supply Chain Content','Newspaper Article',
            'I agree to pay the amount through cheque','Newspaper Article',
           'Magazine','Get updates on DM Content','X Education Forums','Do Not Call'])

leads.drop(del_col, axis = 1, inplace = True)

In [ ]:
leads.shape

In [ ]:
ln = leads.isnull().sum() * 100/len(leads)
ln[ln > 0]

In [ ]:
#dropping columns with high percentage of nulls > 35%
del_col = (['Asymmetrique Activity Index','Asymmetrique Profile Index','Asymmetrique Activity Score',
            'Asymmetrique Profile Score','Lead Quality','Tags'])               
leads.drop(del_col, axis = 1, inplace = True)

In [ ]:
#removing Id columns as not relevant to our analysis
leads.drop(['Prospect ID','Lead Number'], axis = 1, inplace = True)

In [ ]:
#treating nulls
#SUbstitue generic value for null
leads['Lead Source'].value_counts()
leads['Lead Source'].fillna("Other", inplace = True) 
leads['Lead Source'].isnull().sum()

In [ ]:
#least number of total visits possible is 0
leads['TotalVisits'].value_counts()
leads['TotalVisits'].fillna(0, inplace = True) 
leads['TotalVisits'].isnull().sum()

In [ ]:
#Replacing with minimum value possible - 0
leads['Page Views Per Visit'].value_counts()
leads['Page Views Per Visit'].fillna(0, inplace = True) 
leads['Page Views Per Visit'].isnull().sum()

In [ ]:
leads['Last Activity'].value_counts()
leads['Last Activity'].fillna("Other", inplace = True) 
leads['Last Activity'].isnull().sum()

In [ ]:
#
leads['Country'].value_counts()
leads['Country'].fillna("unknown", inplace = True) 
leads['Country'].isnull().sum()

In [ ]:
leads.shape

### Outliers

In [ ]:
lead_n = leads.select_dtypes(include=np.number)
num_col = lead_n.columns
num_col

lead_c = leads.select_dtypes(include=np.object)
cat_col = lead_c.columns
(num_col,cat_col)

In [ ]:
# for i in leads.columns:
plt.figure(figsize = (16,8))
subplot = 221
for i in leads[num_col]:
    plt.subplot(subplot)
    sns.boxplot(leads[i])
    subplot +=1
plt.tight_layout()

In [ ]:
#Outlier treatment

## 2. EDA

In [ ]:
for column in cat_col:
    plt.figure(figsize=(30,6))
    plt.subplot(121)
    sns.countplot(x=leads[column],hue=leads['Converted'],data=leads)
    plt.title(column)    
    plt.xticks(rotation=45)

In [ ]:
#Observations